### ISSUE TITLE
export-file-local-symbols doesn't work when compiling+linking

### ISSUE THREAD

**Issue by karkhaz** [diffblue/cbmc#4689]
The `--export-file-local-symbols` flag appears to have no effect when doing a combined compile+link (i.e., when not using the `-c` flag to goto-cc).

**Comment by karkhaz** [comment]
Here is a more detailed explanation of this bug.

## Context
`goto-cc` has a flag, `--export-file-local-symbols`, that removes the `static` attribute from functions that it is compiling. Consider the following two files:
```c
// main.c
extern int foo();
int main(){
  return foo();
}
```
```c
// foo.c
#include <assert.h>
static int foo(){
  assert(0);
  return 0;
}
```
It is not possible to compile and link these files together correctly, as they are written, because `main` cannot call `foo`---because `foo` is marked as `static`, or file-local. Indeed, if we try to do this, CBMC will not catch the failing assertion, because the body of `foo` is not available.
```sh
goto-cc -c foo.o foo.c
goto-cc -c main.o main.c
goto-cc -o bad-link.goto main.o foo.o
cbmc bad-link.goto
# ...
**** WARNING: no body for function foo
# ...
VERIFICATION successful
```
If we do want to allow `main` to call `foo`, we can use the `--export-file-local-symbols` flag when compiling `foo.c`. This flag has two effects:
* `goto-cc` ignores the `static` attribute on `foo`, meaning that other files can now call it
* `goto-cc` mangles the name of `foo`. Because there might be multiple static functions called `foo` in different files, `goto-cc` changes the name of `foo` so that it is hopefully unique. This means that we need to call the mangled name from `main`.

`foo`'s mangled name, in this case, is `__CPROVER_file_local_foo_c_foo`---that is, the name of the file followed by the name of the function†. So we can try compiling and linking again, ensuring that we use the mangled name from `main`:
```c
// main.c
extern int __CPROVER_file_local_foo_c_foo();

int main(){
  return __CPROVER_file_local_foo_c_foo();
}
```
```c
// foo.c
#include <assert.h>

static int foo()
{
  assert(0);
  return 5;
}
```
```sh
goto-cc --export-file-local-symbols -c foo.o foo.c
goto-cc -c main.o main.c
goto-cc -o good-link.goto main.o foo.o
cbmc good-link.goto
# ...
[foo.assertion.1] line 5 assertion 0: FAILURE

** 1 of 1 failed (2 iterations)
VERIFICATION FAILED
```
## The bug
The above invocations worked as we expected: we're able to link the two files together and call from one into the other. However, if we pass *both* source files into `goto-cc` at once---rather than compiling them separately and then linking them---the `--export-file-local-symbols` flag has no effect.
```sh
goto-cc --export-file-local-symbols -o combined-link.goto main.c foo.c
cbmc combined-link.goto
# ...
**** WARNING: no body for function __CPROVER_file_local_static_c_foo
# ...
VERIFICATION SUCCESSFUL
```
The expected behaviour is that the `--export-file-local-symbols` flag should mangle and export all static functions in all source files when doing a combined compile+link, just like it does when doing only a compile.

† We can find the mangled name of functions by running `goto-instrument --show-symbol-table` on a goto-binary. This will display both the original ('pretty') name and the mangled name for each function.

**Comment by NlightNFotis** [comment]
Hi @karkhaz.

As mentioned before we have found an issue with the implementation of this causing name mangling in `goto-cc` crash with an invariant, as documented in #5380. 

We have a preliminary fix in #5385, but we wanted to make sure that this replicates the behaviour you want. We have changed in that PR the name mangler to be based on hashing instead of the local file name. We want to run it by you to tell us if this change results in a change in the behaviour you expect from the tool.

This no longer crashes `goto-cc` when we run `make goto` in the aws-c-common repository jobs, but it has the unpleasant outcome that it has made testing a lot more difficult (we need to replicate both what happened in the beginning and caused the crash, and the now broken test because while hashing is deterministic in each machine, because it hashes the path, the paths are different between the various machines which complicates the original test case).

Assuming this is more or less what you expect we need to find a way to fix the tests. If not, then we need to do it a different way, so we shouldn't bother with the tests for now.

**Comment by karkhaz** [comment]
@NlightNFotis thanks for looking into this.

I would strongly prefer not to change the implementation of the name-mangling now. Although I included the djb-mangling implementation as an example, I feel that we're basically locked in to the file-mangling implementation, since we use the mangled names all over the place. `git-grep`ing for `__CPROVER_file_local` in -aws-c-common and aws-encryption-sdk-c reveals dozens of uses, for example things like:

```
UNWINDSET +=__CPROVER_file_local_hash_table_c_s_find_entry1.0:$(shell echo $$((1 + $(MAX_TABLE_SIZE))))
UNWINDSET +=__CPROVER_file_local_hash_table_c_s_remove_entry.0:$(shell echo $$((1 + $(MAX_TABLE_SIZE))))
ADDITIONAL_REMOVE_FUNCTION_BODY += --remove-function-body __CPROVER_file_local_hash_table_c_s_emplace_item
ADDITIONAL_REMOVE_FUNCTION_BODY += --remove-function-body __CPROVER_file_local_hash_table_c_s_expand_table
```

I think these would silently cease to have any effect if we changed the name-mangling implementation.

I will look into the bug in #5380 in more detail. I can imagine what's going on, but need to read the code in more detail. I'll write a brief comment on that ticket.
